File: backend/services/news_service.py
"""
news_service.py — Fetches live news from NewsAPI + GNews, merges & deduplicates.
"""
import httpx
from typing import List, Dict
from config import NEWSAPI_KEY, GNEWS_KEY


async def fetch(country_name: str, iso_code: str) -> List[Dict]:
    """Fetch articles from NewsAPI and GNews concurrently, deduplicate by URL."""
    articles: List[Dict] = []
    seen_urls: set = set()
    
    # NewsAPI supported countries for 'country' param
    SUPPORTED_ISO = {
        "ae", "ar", "at", "au", "be", "bg", "br", "ca", "ch", "cn", "co", "cu", "cz", "de", "eg", 
        "fr", "gb", "gr", "hk", "hu", "id", "ie", "il", "in", "it", "jp", "kr", "lt", "lv", "ma", 
        "mx", "my", "ng", "nl", "no", "nz", "ph", "pl", "pt", "ro", "rs", "ru", "sa", "se", "sg", 
        "si", "sk", "th", "tr", "tw", "ua", "us", "ve", "za"
    }

    async with httpx.AsyncClient(timeout=8.0) as client:
        # ── NewsAPI ──────────────────────────────────────────────────────────
        try:
            params = { "apiKey": NEWSAPI_KEY, "pageSize": 10 }
            
            # Use top-headlines if country is supported, otherwise search everything
            if iso_code.lower() in SUPPORTED_ISO:
                url = "https://newsapi.org/v2/top-headlines"
                params["country"] = iso_code.lower()
            else:
                url = "https://newsapi.org/v2/everything"
                params["q"] = country_name
                params["language"] = "en"
                params["sortBy"] = "relevancy"

            resp = await client.get(url, params=params)
            if resp.status_code == 200:
                data = resp.json()
                for a in data.get("articles", []):
                    u = a.get("url", "")
                    if u and u not in seen_urls:
                        seen_urls.add(u)
                        articles.append({
                            "title": a.get("title") or "",
                            "description": a.get("description") or "",
                            "url": u,
                            "source": (a.get("source") or {}).get("name", "NewsAPI"),
                            "publishedAt": a.get("publishedAt") or "",
                        })
            else:
                print(f"[news_service] NewsAPI Error {resp.status_code}: {resp.text}")
        except Exception as e:
            print(f"[news_service] NewsAPI exception: {e}")

        # ── GNews ────────────────────────────────────────────────────────────
        try:
            resp = await client.get(
                "https://gnews.io/api/v4/search",
                params={
                    "q": country_name,
                    "lang": "en",
                    "max": 5,
                    "apikey": GNEWS_KEY,
                },
            )
            if resp.status_code == 200:
                data = resp.json()
                for a in data.get("articles", []):
                    u = a.get("url", "")
                    if u and u not in seen_urls:
                        seen_urls.add(u)
                        articles.append({
                            "title": a.get("title") or "",
                            "description": a.get("description") or "",
                            "url": u,
                            "source": (a.get("source") or {}).get("name", "GNews"),
                            "publishedAt": a.get("publishedAt") or "",
                        })
            else:
                print(f"[news_service] GNews Error {resp.status_code}: {resp.text}")
        except Exception as e:
            print(f"[news_service] GNews exception: {e}")

    return articles

File: backend/services/newsService.js
const axios = require("axios");

const PERSONA_QUERIES = {
    student: "student visa international education university",
    businessman: "economy trade business investment regulation",
    traveler: "tourism safety travel advisory entry requirements",
    remote_worker: "digital nomad visa internet cost of living coworking",
    investor: "stock market real estate FDI interest rates",
};

async function fetchNews(countryName, isoCode, persona = "student", personaDetails = {}) {
    const articles = [];
    const seenUrls = new Set();
    let extraQ = PERSONA_QUERIES[persona] || "";
    
    // Add persona details to query
    if (personaDetails) {
        if (persona === "student" && personaDetails.domain) {
            extraQ += ` ${personaDetails.domain}`;
        } else if (persona === "businessman") {
            if (personaDetails.domain) extraQ += ` ${personaDetails.domain}`;
            if (personaDetails.focus) extraQ += ` ${personaDetails.focus}`;
        } else if (persona === "traveler") {
            if (personaDetails.interests) extraQ += ` ${personaDetails.interests}`;
            if (personaDetails.season) extraQ += ` ${personaDetails.season}`;
        } else if (persona === "remote_worker" && personaDetails.industry) {
            extraQ += ` ${personaDetails.industry}`;
        } else if (persona === "investor" && personaDetails.asset) {
            extraQ += ` ${personaDetails.asset}`;
        }
    }

    const query = `${countryName} ${extraQ}`.trim();

    // NewsAPI
    try {
        const newsApiResp = await axios.get("https://newsapi.org/v2/everything", {
            params: {
                apiKey: process.env.NEWSAPI_KEY,
                q: query,
                language: "en",
                sortBy: "publishedAt",
                pageSize: 10
            }
        });

        if (newsApiResp.data && newsApiResp.data.articles) {
            newsApiResp.data.articles.forEach(a => {
                if (a.url && !seenUrls.has(a.url)) {
                    seenUrls.add(a.url);
                    articles.push({
                        title: a.title || "",
                        description: a.description || "",
                        url: a.url,
                        source: a.source ? a.source.name : "NewsAPI",
                        publishedAt: a.publishedAt || ""
                    });
                }
            });
        }
    } catch (err) {
        console.error("[newsService] NewsAPI error:", err.message);
    }

    // GNews
    try {
        const gnewsResp = await axios.get("https://gnews.io/api/v4/search", {
            params: {
                q: query,
                lang: "en",
                max: 5,
                apikey: process.env.GNEWS_KEY
            }
        });

        if (gnewsResp.data && gnewsResp.data.articles) {
            gnewsResp.data.articles.forEach(a => {
                if (a.url && !seenUrls.has(a.url)) {
                    seenUrls.add(a.url);
                    articles.push({
                        title: a.title || "",
                        description: a.description || "",
                        url: a.url,
                        source: a.source ? a.source.name : "GNews",
                        publishedAt: a.publishedAt || ""
                    });
                }
            });
        }
    } catch (err) {
        console.error("[newsService] GNews error:", err.message);
    }

    return articles;
}

module.exports = { fetchNews };
